### PR TITLE
osd/OSDMap: improve the performance of pg_to_acting_osds

### PR DESCRIPTION
**Pull request by liupan1111**
…up vector" is not requested.

Signed-off-by: Pan Liu <pan.liu@istuary.com>

### CODE REVIEW

**Comment by liupan1111** [comment]
@liewegas @tchaikov , please help take a look.

**Comment by liupan1111** [comment]
The case is: when pg temp is not empty, acting will be calculated without ups and up_primary. But ups and up_primary are always calculated, and affect the performance at this situation.

**Comment by liupan1111** [comment]
@liewegas @tchaikov , please help take a look.

**Comment by liupan1111** [comment]
I find one thing very weird. I created this pr yesterday morning, and found one testcase failed when ran jekins: 
qa/workunits/cephtool/test.sh : test_mon_crushmap_validation
"Error EINVAL: Parse error setting osd_scrub_auto_repair_num_errors to '-1' using injectargs."

So I had to close this pr, and investigated what was wrong, but nothing found. When I reopened this pr without any changes, jekins passed....

Anyone could give me any idea?

**Comment by liewegas** [comment]
@liupan1111 there are still some issues with the unit tests; it sounds like it wasn't caused by this change.

File: src/osd/OSDMap.cc [hunk 1]
```diff
@@ -1720,14 +1720,16 @@ void OSDMap::_pg_to_up_acting_osds(const pg_t& pg, vector<int> *up, int *up_prim
...
-  if (_acting.empty()) {
-    _acting = _up;
-    if (_acting_primary == -1) {
-      _acting_primary = _up_primary;
+  if (_acting.empty() || up || up_primary) {
+    _pg_to_raw_osds(*pool, pg, &raw, &_up_primary, &pps);
+    _raw_to_up_osds(*pool, raw, &_up, &_up_primary);
+    _apply_primary_affinity(pps, *pool, &_up, &_up_primary);
+    if (_acting.empty()) {
+	  _acting = _up;
```
Comment (liewegas) [inline src/osd/OSDMap.cc]: indentation?

File: src/osd/OSDMap.cc [hunk 2]
```diff
@@ -1720,14 +1720,16 @@ void OSDMap::_pg_to_up_acting_osds(const pg_t& pg, vector<int> *up, int *up_prim
...
+      if (_acting_primary == -1) {
+        _acting_primary = _up_primary;
+      }
     }
   }
   if (up)
```
Comment (liewegas) [inline src/osd/OSDMap.cc]: you could mvoe the if (up) and if (up_prmiary) lines into the block?

**Comment by liupan1111** [comment]
@liewegas , modified as your request.

File: src/osd/OSDMap.cc (hunk 1 above)
Comment (tchaikov) [inline src/osd/OSDMap.cc]: @liupan1111 ping?

File: src/osd/OSDMap.cc [hunk 3]
```diff
@@ -1720,20 +1720,24 @@ void OSDMap::_pg_to_up_acting_osds(const pg_t& pg, vector<int> *up, int *up_prim
...
-  if (_acting.empty()) {
-    _acting = _up;
-    if (_acting_primary == -1) {
-      _acting_primary = _up_primary;
+  if (_acting.empty() || up || up_primary) {
+    _pg_to_raw_osds(*pool, pg, &raw, &_up_primary, &pps);
+    _raw_to_up_osds(*pool, raw, &_up, &_up_primary);
+    _apply_primary_affinity(pps, *pool, &_up, &_up_primary);
+    if (_acting.empty()) {
+        _acting = _up;
```
Comment (tchaikov) [inline src/osd/OSDMap.cc]: @liupan1111 could you fix the indent?

File: src/osd/OSDMap.cc (hunk 1 above)
Comment (tchaikov) [inline src/osd/OSDMap.cc]: and could you fix the Signed-off-by line in your commit message? or just squash these two commits into a single one?

**Comment by liupan1111** [comment]
@tchaikov  done, thanks!

**Comment by tchaikov** [comment]
@liupan1111 could you fix the signed-off-by line in 993d37d, or better off squashing these two commit into a single one?

**Comment by liupan1111** [comment]
@tchaikov what is wrong with signed-off-by?

**Comment by tchaikov** [comment]
```diff
- Signed-off-by: Pan Liu pan.liu@istuary.com
+ Signed-off-by: Pan Liu <pan.liu@istuary.com>
```

**Comment by liupan1111** [comment]
@tchaikov modified.

**Comment by tchaikov** [comment]
http://pulpito.ceph.com/kchai-2016-12-10_12:27:10-rados-wip-kefu-testing---basic-mira/

**Pull request merged**